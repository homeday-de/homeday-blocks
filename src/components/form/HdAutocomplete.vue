<template>
  <div
    class="autocomplete"
    @keydown.down="onDown"
    @keydown.up="onUp"
    @keydown.enter="onEnter"
    @keydown.esc="hide"
  >
    <hd-input v-bind="inputProps" :value="query" @dataChange="updateQuery" class="autocomplete__field" />
    <ul class="autocomplete__suggestions" v-if="suggestions.length > 0 && !suggestionsHidden">
      <li
        class="autocomplete__suggestionItem" @click="selectSuggestion(suggestion, true)" v-for="(suggestion, index) in suggestions" :key="index">
        <slot :suggestion="suggestion" :isSelected="suggestion === selectedSuggestion"></slot>
      </li>
    </ul>
  </div>
</template>

<script>
import HdInput from 'hd-blocks/components/form/HdInput.vue';
import { populateTemplate } from 'hd-blocks/services/utils';
// import debounce from 'lodash/debounce';

export default {
  name: 'HdAutocomplete',
  components: {
    HdInput,
  },
  props: {
    formatter: {
      type: String,
      required: true,
    },
    suggestions: {
      type: Array,
    },
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      default: 'text',
    },
    placeholder: {
      type: String,
      required: true,
    },
    required: {
      type: Boolean,
      default: false,
    },
    autocomplete: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    lang: {
      type: String,
      default: 'de',
      validator: val => ['de', 'en'].indexOf(val) !== -1,
    },
  },
  data() {
    return {
      query: '',
      queryBackup: '',
      selectedSuggestion: null,
      suggestionsHidden: false,
    };
  },
  methods: {
    onDown() {
      const noSelection = this.selectedSuggestion === null;
      const isLast = (this.nextSuggestionIndex + 1) > this.suggestions.length;
      if (noSelection) {
        this.selectSuggestion(this.suggestions[0]);
      } else if (isLast) {
        this.query = this.queryBackup;
        this.selectSuggestion(null);
      } else {
        this.selectSuggestion(this.suggestions[this.nextSuggestionIndex]);
      }
    },
    onUp() {
      const noSelection = this.selectedSuggestion === null;
      const isFirst = this.previousSuggestionIndex === -1;
      if (noSelection) {
        this.selectSuggestion(this.suggestions[this.suggestions.length - 1]);
      } else if (isFirst) {
        this.query = this.queryBackup;
        this.selectSuggestion(null);
      } else {
        this.selectSuggestion(this.suggestions[this.previousSuggestionIndex]);
      }
    },
    hide() {
      this.suggestionsHidden = true;
    },
    updateQuery({ value }) {
      this.query = value;
    },
    formatQuery(suggestion) {
      const formatVars = this.formatter.match(/:(.+?)\b/g);
      const formatVarsTrimmed = formatVars.map(v => v.replace(/:/g, ''));
      const formatterValues = formatVarsTrimmed.reduce((acc, curr) => ({
        ...acc, [curr]: suggestion[curr],
      }), {});

      return populateTemplate(this.formatter, formatterValues);
    },
    selectSuggestion(suggestion, hideList) {
      if (this.queryBackup === '') {
        this.queryBackup = this.query;
      }
      if (suggestion) {
        this.query = this.formatQuery(suggestion);
      } else {
        this.queryBackup = '';
      }
      this.selectedSuggestion = suggestion;

      if (hideList) {
        this.hide();
      }
    },
  },
  computed: {
    currentSuggestionIndex() {
      return this.suggestions.findIndex(s => s === this.selectedSuggestion);
    },
    previousSuggestionIndex() {
      return this.currentSuggestionIndex - 1;
    },
    nextSuggestionIndex() {
      return this.currentSuggestionIndex + 1;
    },
    inputProps() {
      const {
        name, label, type, placeholder, required, autocomplete, autofocus, lang,
      } = this;
      return {
        name, label, type, placeholder, required, autocomplete, autofocus, lang,
      };
    },
  },
  watch: {
    query(val) {
      this.$emit('queryUpdated', val);
      this.suggestionsHidden = false;
    },
    suggestions() {
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'hd-blocks/styles/mixins.scss';

.autocomplete {
  &__suggestions {
    overflow: hidden;
    box-shadow: 0 2px 4px rgba(0,0,0,0.2);
  }
  &__suggestionItem {
    list-style: none;
  }
  &__field {
    margin-bottom: 0;
  }
}
</style>
