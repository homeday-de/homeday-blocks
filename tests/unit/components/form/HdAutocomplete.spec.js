import { shallowMount } from '@vue/test-utils';
import HdAutocomplete from '@/components/form/HdAutocomplete.vue';

describe('HdAutocomplete', () => {
  let wrapper;
  const SUGGESTIONS_LOCATOR = '.autocomplete__suggestions';
  const OPTIONS = ['Option 1', 'Option 2', 'Option 3'];

  beforeEach(() => {
    wrapper = shallowMount(HdAutocomplete, {
      propsData: {
        formatter: ':bold',
        suggestions: OPTIONS,
        name: 'Test',
        label: 'Test',
        type: 'text',
        placeholder: 'Test',
        required: false,
        autocomplete: 'on',
        autofocus: false,
        lang: 'en',
      },
    });
  });

  test('the component is rendered', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('the lang prop validation accepts only danish and english', () => {
    expect(wrapper.vm.$options.props.lang.validator('pt')).toBeFalsy();
    expect(wrapper.vm.$options.props.lang.validator('en')).toBeTruthy();
    expect(wrapper.vm.$options.props.lang.validator('de')).toBeTruthy();
  });

  test('Query prop update emits event and show options on view', () => {
    // Check for suggestions not displayed
    expect(wrapper.emitted('queryUpdated')).toBeFalsy();

    // Change query to check suggestions displayed
    wrapper.vm.updateQuery({ value: 'Option' });
    expect(wrapper.emitted('queryUpdated')).toBeTruthy();
    expect(wrapper.find(SUGGESTIONS_LOCATOR).element.childElementCount).toBe(3);
  });

  test('esc key should hide suggestions', () => {
    // Watch the hide function
    const spyHide = jest.spyOn(wrapper.vm, 'hide');

    // Change query to check suggestions displayed
    wrapper.vm.updateQuery({ value: 'Option' });
    expect(wrapper.find(SUGGESTIONS_LOCATOR).element.childElementCount).toBe(3);

    wrapper.trigger('keydown.esc');
    expect(wrapper.vm.$data.suggestionsHidden).toBeTruthy();
    expect(spyHide).toBeCalled();
  });

  test('keydown.down should switch suggestion focus to next on list', () => {
    // Change query to check suggestions displayed
    wrapper.vm.updateQuery({ value: 'Option' });

    wrapper.trigger('keydown.down');
    expect(wrapper.vm.$data.selectedSuggestion).toBe(OPTIONS[0]);
    expect(wrapper.vm.currentSuggestionIndex).toBe(0);

    wrapper.trigger('keydown.down');
    expect(wrapper.vm.$data.selectedSuggestion).toBe(OPTIONS[1]);
    expect(wrapper.vm.currentSuggestionIndex).toBe(1);

    wrapper.trigger('keydown.down');
    expect(wrapper.vm.$data.selectedSuggestion).toBe(OPTIONS[2]);
    expect(wrapper.vm.currentSuggestionIndex).toBe(2);

    wrapper.trigger('keydown.down');
    expect(wrapper.vm.$data.selectedSuggestion).toBeNull();
  });

  test('keydown.up should switch suggestion focus to previous on list', () => {
    // Change query to check suggestions displayed
    wrapper.vm.updateQuery({ value: 'Option' });

    wrapper.trigger('keydown.up');
    expect(wrapper.vm.$data.selectedSuggestion).toBe(OPTIONS[2]);
    expect(wrapper.vm.currentSuggestionIndex).toBe(2);

    wrapper.trigger('keydown.up');
    expect(wrapper.vm.$data.selectedSuggestion).toBe(OPTIONS[1]);
    expect(wrapper.vm.currentSuggestionIndex).toBe(1);

    wrapper.trigger('keydown.up');
    expect(wrapper.vm.$data.selectedSuggestion).toBe(OPTIONS[0]);
    expect(wrapper.vm.currentSuggestionIndex).toBe(0);

    wrapper.trigger('keydown.up');
    expect(wrapper.vm.$data.selectedSuggestion).toBeNull();
  });
});
