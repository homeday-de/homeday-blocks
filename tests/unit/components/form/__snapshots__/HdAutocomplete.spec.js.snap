// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdAutocomplete the component is rendered 1`] = `
<div class="autocomplete">
  <hdinput-stub label="Test" name="Test" type="text" value="" placeholder="Test" autocomplete="on" lang="en" texts="[object Object]" icon="" class="autocomplete__field"></hdinput-stub>
  <ul class="autocomplete__suggestions">
    <li class="autocomplete__suggestionItem"></li>
    <li class="autocomplete__suggestionItem"></li>
    <li class="autocomplete__suggestionItem"></li>
  </ul>
</div>
`;
